x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/highlight/SemanticTextHighlighter.java

```diff
@@ -105,28 +115,36 @@ public HighlightField highlight(FieldHighlightContext fieldContext) throws IOExc
         int size = Math.min(chunks.size(), numberOfFragments);
         if (fieldContext.field.fieldOptions().scoreOrdered() == false) {
             chunks = chunks.subList(0, size);
-            chunks.sort(Comparator.comparingInt(c -> c.offset));
+            chunks.sort(Comparator.comparingInt(c -> c.index));
         }
         Text[] snippets = new Text[size];
-        List<Map<?, ?>> nestedSources = XContentMapValues.extractNestedSources(
-            fieldType.getChunksField().fullPath(),
-            fieldContext.hitContext.source().source()
-        );
+        final Function<OffsetAndScore, String> offsetToContent;
+        if (fieldType.useLegacyFormat()) {
+            List<Map<?, ?>> nestedSources = XContentMapValues.extractNestedSources(
+                fieldType.getChunksField().fullPath(),
+                fieldContext.hitContext.source().source()
+            );
+            offsetToContent = entry -> getContentFromNestedSourcesLegacy(fieldType.name(), entry, nestedSources);
+        } else {
+            Map<String, String> fieldToContent = new HashMap<>();
+            offsetToContent = entry -> {
+                String content = fieldToContent.computeIfAbsent(entry.offset().field(), key -> {
+                    try {
+                        return extractFieldContent(
+                            fieldContext.context.getSearchExecutionContext(),
+                            fieldContext.hitContext,
+                            entry.offset.field()
+                        );
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
```

Comment:
Nitpick: Add a message to the new Exception. 

```suggestion
                        throw new UncheckedIOException("Error extracting field content from field " + entry.offset.field(), e);
```